# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webauthn/WebAuthnManager.cpp
# Commit: d28545793e92
# Full Hash: d28545793e92c94cf8355690574ea578e3e8b1c9
# Author: J.C. Jones <jjones@mozilla.com>
# Date: 2019-04-02 08:35:12
# Regressor Bug: 1448408
# File Overlap Count: 2
# Description:
#   Bug 1540378 - Web Authentication: Fix teardown during cycle collection r=keeler,mccr8
#   
#   In Bug 1448408 ("Don't listen to visibility events"), it became possible to
#   close a tab without a visibility event to cause transactions to cancel. This
#   is a longstanding bug that was covered up by the visibility events. This patch
# ==============================================================================

diff -r dccd58970fb0 -r d28545793e92 dom/webauthn/WebAuthnManager.cpp
--- a/dom/webauthn/WebAuthnManager.cpp	Tue Apr 02 00:30:57 2019 +0000
+++ b/dom/webauthn/WebAuthnManager.cpp	Mon Apr 01 23:13:26 2019 +0000
@@ -38,8 +38,18 @@
 NS_IMPL_ISUPPORTS_CYCLE_COLLECTION_INHERITED_0(WebAuthnManager,
                                                WebAuthnManagerBase)
 
-NS_IMPL_CYCLE_COLLECTION_INHERITED(WebAuthnManager, WebAuthnManagerBase,
-                                   mFollowingSignal, mTransaction)
+NS_IMPL_CYCLE_COLLECTION_CLASS(WebAuthnManager)
+NS_IMPL_CYCLE_COLLECTION_UNLINK_BEGIN_INHERITED(WebAuthnManager,
+                                                WebAuthnManagerBase)
+  NS_IMPL_CYCLE_COLLECTION_UNLINK(mFollowingSignal)
+  NS_IMPL_CYCLE_COLLECTION_UNLINK(mTransaction)
+  tmp->ClearTransaction();
+NS_IMPL_CYCLE_COLLECTION_UNLINK_END
+NS_IMPL_CYCLE_COLLECTION_TRAVERSE_BEGIN_INHERITED(WebAuthnManager,
+                                                  WebAuthnManagerBase)
+  NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mFollowingSignal)
+  NS_IMPL_CYCLE_COLLECTION_TRAVERSE(mTransaction)
+NS_IMPL_CYCLE_COLLECTION_TRAVERSE_END
 
 /***********************************************************************
  * Utility Functions
@@ -156,7 +166,7 @@
  **********************************************************************/
 
 void WebAuthnManager::ClearTransaction() {
-  if (!NS_WARN_IF(mTransaction.isNothing())) {
+  if (!mTransaction.isNothing()) {
     StopListeningForVisibilityEvents();
   }
 